File: codejam/cpp/2009/1A/1A_CrossingTheRoad.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<ll> vi;
typedef pair<ll,ll> pi;
#define FOR(i,a) for (ll i = 0; i < (a); i++)
#define len(x) (ll) x.size()
const ll INF = 1LL << 62;
const ll MOD = 1000000007;
//const ll MOD = 998244353;
const double PI = 4*atan(double(1.0));
vector<vector<ll>> twodi(ll N, ll M, ll v) { vector<vector<ll>> res(N); for (auto &vv : res) vv.resize(M,v); return res; }

struct dnode { ll d,i,j; };
bool gt(dnode a, dnode b) { return a.d > b.d; } // stl crap
int main (int argc, char **argv) {
    if (argc > 1) { freopen(argv[1],"r",stdin); }
    ios_base::sync_with_stdio(false);cin.tie(0);
    // PROGRAM STARTS HERE
    ll T; cin >> T;
    for (ll tt=1;tt<=T;tt++) {
        ll N,M; cin >> N >> M;
        auto S = twodi(N,M,0); auto W = twodi(N,M,0); auto T = twodi(N,M,0);
        FOR(i,N) FOR(j,M) cin >> S[i][j] >> W[i][j] >> T[i][j];
        ll inf = 1LL << 62;
        auto darr = twodi(2*N,2*M,inf);
        priority_queue<dnode,vector<dnode>,function<bool(dnode,dnode)>> mh(gt);
        mh.push({0,2*N-1,0});
        while (!mh.empty()) {
            auto d = mh.top().d; auto i = mh.top().i; auto j = mh.top().j; mh.pop();
            if (darr[i][j] != inf) continue;
            darr[i][j] = d; auto s = S[i/2][j/2]; auto w = W[i/2][j/2]; auto t = T[i/2][j/2];
            t = t % (s+w) - (s+w);
            auto nstime = (d-t) % (s+w) < s  ? d+1 : t + (d-t)/(s+w)*(s+w) + (s+w) + 1;
            auto ewtime = (d-t) % (s+w) >= s ? d+1 : t + (d-t)/(s+w)*(s+w) + s + 1;
            if ((i&1) == 0) { if (i-1 >= 0) mh.push({d+2,i-1,j}); mh.push({nstime,i+1,j});  }
            if ((i&1) == 1) { mh.push({nstime,i-1,j}); if (i+1 < 2*N) mh.push({d+2,i+1,j}); }
            if ((j&1) == 0) { if (j-1 >= 0) mh.push({d+2,i,j-1}); mh.push({ewtime,i,j+1});  }
            if ((j&1) == 1) { mh.push({ewtime,i,j-1}); if (j+1 < 2*M) mh.push({d+2,i,j+1}); }
        }
        printf("Case #%lld: %lld\n",tt,darr[0][2*M-1]);
    }
}

File: codejam/python/2009/1A/1A_CrossingTheRoad.py
import sys
from collections import deque

## Input crap
infile = sys.stdin
intokens = deque()
def getTokens(): 
    while not intokens:
        for c in infile.readline().rstrip().split() :
            intokens.append(c)    
def gs(): getTokens(); return intokens.popleft()
def gi(): return int(gs())
def gf(): return float(gs())
def gbs(): return [c for c in gs()]
def gis(n): return [gi() for i in range(n)]
def ia(m): return [0] * m
def iai(m,v): return [v] * m
def twodi(n,m,v): return [iai(m,v) for i in range(n)]
def fill2(m) : r = gis(2*m); return r[0::2],r[1::2]
def fill3(m) : r = gis(3*m); return r[0::3],r[1::3],r[2::3]
def fill4(m) : r = gis(4*m); return r[0::4],r[1::4],r[2::4],r[3::4]
MOD = 998244353
INF = 1 << 62
##MOD = 1000000007

def minheaplt (a,b) : return a < b
class minheap :
    def __init__(self,lt=minheaplt) :
        self.lt = lt
        self.h  = []
    def isEmpty(self) : return len(self.h) == 0
    def clear(self) : self.h = self.h[:0]
    def len(self) : return len(self.h)
    def push(self,v) : self.h.append(v); self._siftdown(0,len(self.h)-1)
    def head(self) : return self.h[0]
    def pop(self) : 
        v1 = self.h[0]; l = len(self.h)
        if l == 1 : self.h.pop(); return v1
        l-=1; self.h[0] = self.h[l]; self.h.pop(); self._siftup(0); return v1
    def heapify(self,v) :
        for vv in v : self.h.append(v)
        n = len(self.h)
        for i in range(n//2-1,-1,-1) : self._siftup(i)
    def _siftdown(self,startpos,pos) :
        newitem = self.h[pos]
        while pos > startpos :
            ppos = (pos-1)>>1; p = self.h[ppos]
            if not self.lt(newitem,p) : break
            self.h[pos],pos = p,ppos
        self.h[pos] = newitem
    def _siftup(self,pos) :
        endpos,startpos,newitem,chpos = len(self.h),pos,self.h[pos],2*pos+1
        while chpos < endpos :
            rtpos = chpos+1
            if rtpos < endpos and not self.lt(self.h[chpos],self.h[rtpos]) : chpos = rtpos
            self.h[pos],pos = self.h[chpos],chpos; chpos = 2*pos+1
        self.h[pos] = newitem; self._siftdown(startpos,pos) 

def main() :
    if len(sys.argv) > 1 : global infile; infile = open(sys.argv[1],'rt')
    ## PROGRAM STARTS HERE
    T = gi()
    for tt in range(1,T+1) :
        N,M = gi(),gi()
        S,W,T = [],[],[]
        for i in range(N) : r = gis(3*M); S.append(r[0::3]); W.append(r[1::3]); T.append(r[2::3])
        darr = [[INF]*(2*M) for _ in range(2*N)]
        mh = minheap()
        mh.push((0,2*N-1,0))
        while not mh.isEmpty() :
            (d,i,j) = mh.pop()
            if darr[i][j] != INF : continue
            darr[i][j] = d
            s,w,t = S[i//2][j//2],W[i//2][j//2],T[i//2][j//2]
            t = t % (s+w) - (s+w)
            nstime = d+1 if (d-t) % (s+w) < s else t + (d-t)//(s+w)*(s+w) + (s+w) + 1
            ewtime = d+1 if (d-t) % (s+w) >= s else t + (d-t)//(s+w)*(s+w) + s + 1
            if i&1 == 0 : ## South side of intersection
                if i-1 >= 0 : mh.push((d+2,i-1,j))
                mh.push((nstime,i+1,j))
            else :  ## North side of intersection
                mh.push((nstime,i-1,j))
                if i+1 < 2*N : mh.push((d+2,i+1,j))
            if j&1 == 0 : ## West side of intersection
                if j-1 >= 0 : mh.push((d+2,i,j-1))
                mh.push((ewtime,i,j+1))
            else : ## East side of intersection
                mh.push((ewtime,i,j-1))
                if j+1 < 2*M : mh.push((d+2,i,j+1))
        print(f"Case #{tt}: {darr[0][2*M-1]}")

if __name__ == "__main__" :
    main()

